### PR TITLE
106 create adminPanel, proper display and controllers, implement CMD of orgs and employers

### PR DESCRIPTION
**Pull request by WojciechPtas**
PLEASE REVIEW

### CODE REVIEW

File: controllers/adminController.js [hunk 1]
```diff
@@ -2,35 +2,166 @@ const Employer = require('../models/employerModel');
...
+    }
 }
-const addEmployer_put = async (req, res) => {
+const addEmployer_put  =  async(req, res) =>{
+    const { email, password, name, surname, permissionLevel, organisation_id } = req.body;
 
-}
+    try{
+        const employer = await Employer.create({ email, password, name, surname,permissionLevel,organisation_id });
+        console.log("New employer %s %s with permission level %s created", name, surname , permissionLevel);
```
Comment (WojciechPtas) [inline controllers/adminController.js:67]: space before comma

**Comment by WojciechPtas** [comment]
Middleware is to be implemented

File: controllers/adminController.js [hunk 2]
```diff
@@ -2,35 +2,166 @@ const Employer = require('../models/employerModel');
...
+    try {
+        const employer_id = req.query._id;
+        const all_organisations = (await Organisation.find({})).map(org => org.toObject());
+        if (employer_id) {
+            const employer = (await Employer.findById(employer_id)).toObject();
+            const organisation = (await Organisation.findById(employer.organisation_id)).toObject();
+            res.render('employer/manage_employer', { title: 'Edycja konta pracowniczego', employer, organisation,all_organisations, user: req.session.employer, scrollable: true });
+        } else {
+            const passwordGenerator = ()=>{
+                var length = 8,
```
Comment (macibol775) [inline controllers/adminController.js:14]: var?

Comment (macibol775) [reply controllers/adminController.js:14]: const

File: controllers/adminController.js [hunk 3]
```diff
@@ -2,35 +2,166 @@ const Employer = require('../models/employerModel');
...
+                charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*()_+?><:{}[]";
```
Comment (macibol775) [inline controllers/adminController.js:15]: const

File: controllers/adminController.js [hunk 4]
```diff
@@ -2,35 +2,166 @@ const Employer = require('../models/employerModel');
...
+                var retVal = "";
```
Comment (macibol775) [inline controllers/adminController.js:16]: let

File: controllers/adminController.js [hunk 5]
```diff
@@ -2,35 +2,166 @@ const Employer = require('../models/employerModel');
...
+    } catch (e) {
+        console.log(e);
+        res.sendStatus(500);
+    }
 }
 const addOrganisation_put = async (req, res) => {
+    const { name, description } = req.body;
 
+    try {
+        const org = await Organisation.create({ name, description });
```
Comment (macibol775) [inline controllers/adminController.js:47]: not used

File: controllers/adminController.js [hunk 6]
```diff
@@ -2,35 +2,166 @@ const Employer = require('../models/employerModel');
...
+        res.json({ errors });
+    }
 }
-const addEmployer_put = async (req, res) => {
+const addEmployer_put  =  async(req, res) =>{
+    const { email, password, name, surname, permissionLevel, organisation_id } = req.body;
 
-}
+    try{
+        const employer = await Employer.create({ email, password, name, surname,permissionLevel,organisation_id });
```
Comment (macibol775) [inline controllers/adminController.js:66]: not used

File: controllers/adminController.js [hunk 7]
```diff
@@ -2,35 +2,166 @@ const Employer = require('../models/employerModel');
...
+        res.json({ errors });
+    }
 
 }
 
 const modifyEmployer_post = async (req, res) => {
-
+    const {name, surname, email, permissionLevel, organisation_id, employer_id} = req.body;
+    try{
+        await Employer.findByIdAndUpdate(employer_id, {name, surname, email, permissionLevel,organisation_id});
```
Comment (macibol775) [inline controllers/adminController.js:108]: Czy organisation_id może być modyfikowane?

Comment (macibol775) [inline controllers/adminController.js:108]: brak spacji po przecinku

File: models/employerModel.js [hunk 8]
```diff
@@ -29,6 +29,11 @@ const employerSchema = new mongoose.Schema({
         validate: [isEmail, "Email invalid"],
         unique: [true, "Email already in use"],
     },
+    organisation_id:{
+        type: mongoose.Schema.Types.ObjectId,
```
Comment (macibol775) [inline models/employerModel.js:33]: "mongoose." not needed, copilot detected

File: public/js/modifyEmployer.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
+function initializeModifyEmployerSubmitListener (employer_id){
+    const form = document.getElementById('employerForm');
+    const submitButton = document.getElementById('submitBtn');
+
+    submitButton.addEventListener('click', (event) => {
+        event.preventDefault();
+        const request = new XMLHttpRequest();
+        const formData = new FormData(form);
+        request.onload = () => {
+            console.log(request.response);
```
Comment (macibol775) [inline public/js/modifyEmployer.js:10]: have you found any practical usage of this log?

File: controllers/adminController.js [hunk 10]
```diff
@@ -2,35 +2,166 @@ const Employer = require('../models/employerModel');
...
+        res.sendStatus(200);
+    } catch (e) {
+        console.log(e);
+        res.sendStatus(500);
+    }
 }
 
-const organisation_delete = async (req, res) => {
-
+const delete_organisation_post = async (req, res) => {
```
Comment (macibol775) [inline controllers/adminController.js:136]: deleteOrganisation_post

File: views/organisation/manage_organisation.ejs [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+                <div class="col-sm-offset-2 col-sm-12">
+                    <button type="submit" class="btn" id="submitBtn">Opublikuj</button>
+                </div>
+            </form>
+        </div>
+    </div>
+</div>
+
+<script type="module">
+    <% var organisation_id = locals.organisation ? organisation._id : null;%>
```
Comment (MasterCoookie) [inline views/organisation/manage_organisation.ejs:32]: dont use var

**Review by MasterCoookie — CHANGES_REQUESTED** [review]
Change vars and protect apropriate routes

File: views/employer/manage_employer.ejs [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+                    <label class="control-label col-sm-8 pt-2 pb-1 m-0" for="name">Imię:</label>
+                    <div class="col-sm-8 mx-auto">
+                        <input type="text" name="name" id="name" value="<% if(locals.employer && employer.name != null) {%><%= employer.name %><% } %>" required="" class="form-control shadow bg-white rounded"/><br/>
+                    </div>
+
+                    <label class="control-label col-sm-8 pt-2 pb-1 m-0" for="surname">Nazwisko:</label>
+                    <div class="col-sm-8 mx-auto">
+                        <input type="text" id="surname" name="surname" rows="8" cols="64" class="form-control shadow bg-white rounded" value="<% if(locals.employer && employer.surname != null) {%><%= employer.surname %><% } %>" required="" class="form-control shadow bg-white rounded"/><br/>
+                    </div>
+                    <label class="control-label col-sm-8 pt-2 pb-1 m-0" for="surname">Email:</label>
```
Comment (macibol775) [inline views/employer/manage_employer.ejs:19]: for=dupa

File: views/employer/manage_employer.ejs [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+                    <div class="col-sm-8 mx-auto">
+                        <input type="text" id="email" name="email" rows="8" cols="64" class="form-control shadow bg-white rounded" value="<% if(locals.employer && employer.email != null) {%><%= employer.email %><% } %>" required="" class="form-control shadow bg-white rounded"/><br/>
```
Comment (macibol775) [inline views/employer/manage_employer.ejs:21]: type=email

File: views/employer/manage_employer.ejs [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+                        <select type="text" id="organisation_id" name="organisation_id" rows="8" cols="64" class="form-control shadow bg-white rounded" value="<% if(locals.organisation && organisation.name != null) {%><%= organisation.name %><% } %>" required="" class="form-control shadow bg-white rounded">
+                            <%if (all_organisations && all_organisations.length > 0){ 
+                                all_organisations.forEach( org => { %>
+                            <option value="<%=org._id%>"><%=org.name%></option>
+                            <%})}%>
+                        </select><br/>
+                    </div>
+                    <label class="control-label col-sm-8 pt-2 pb-1 m-0" for="permissionLevel">Poziom dostępu:</label>
+                    <div class="col-sm-8 mx-auto">
+                        <select type="text" id="permissionLevel" name="permissionLevel" rows="8" cols="64" class="form-control shadow bg-white rounded" value="<% if(locals.employer && employer.permissionLevel != null) {%><%= employer.permissionLevel %><% } %>" required="" class="form-control shadow bg-white rounded">
```
Comment (macibol775) [inline views/employer/manage_employer.ejs:34]: select does not have type attribute (https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select#attributes)

File: views/employer/manage_employer.ejs [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+                    </div>
+                    <label class="control-label col-sm-8 pt-2 pb-1 m-0" for="surname">Email:</label>
+                    <div class="col-sm-8 mx-auto">
+                        <input type="text" id="email" name="email" rows="8" cols="64" class="form-control shadow bg-white rounded" value="<% if(locals.employer && employer.email != null) {%><%= employer.email %><% } %>" required="" class="form-control shadow bg-white rounded"/><br/>
+                    </div>
+                    <label class="control-label col-sm-8 pt-2 pb-1 m-0" for="organisation_id">Organizacja:</label>
+                    <div class="col-sm-8 mx-auto">
+                        <select type="text" id="organisation_id" name="organisation_id" rows="8" cols="64" class="form-control shadow bg-white rounded" value="<% if(locals.organisation && organisation.name != null) {%><%= organisation.name %><% } %>" required="" class="form-control shadow bg-white rounded">
```
Comment (macibol775) [inline views/employer/manage_employer.ejs:25]: jak niżej

Comment (macibol775) [reply views/employer/manage_employer.ejs:25]: rows, cols?

File: views/employer/manage_employer.ejs (hunk 14 above)
Comment (macibol775) [reply views/employer/manage_employer.ejs:34]: rows, cols?

File: views/employer/manage_employer.ejs [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+                        <select type="text" id="permissionLevel" name="permissionLevel" rows="8" cols="64" class="form-control shadow bg-white rounded" value="<% if(locals.employer && employer.permissionLevel != null) {%><%= employer.permissionLevel %><% } %>" required="" class="form-control shadow bg-white rounded">
+                            <option value="1">Sekretarka</option>
+                            <option value="2">Rekruter</option>
+                            <option value="3">Admin</option>
+                        </select><br/>
+                    </div>
+                    <%if(locals.passwordGenerator){%>
+                    <label class="control-label col-sm-8 pt-2 pb-1 m-0" for="password">Hasło:</label>
+                    <div class="col-sm-8 mx-auto">
+                        <input type="text" id="surname" name="password" rows="8" cols="64" class="form-control shadow bg-white rounded" value="<%=passwordGenerator()%>" required="" class="form-control shadow bg-white rounded"/><br/>
```
Comment (macibol775) [inline views/employer/manage_employer.ejs:43]: id="surname" name="password"
rows cols???!!!???

File: views/employer/manage_employer.ejs [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+            <form class="form-horizontal" id="employerForm">
+                <div class="form-group">
+                    <label class="control-label col-sm-8 pt-2 pb-1 m-0" for="name">Imię:</label>
+                    <div class="col-sm-8 mx-auto">
+                        <input type="text" name="name" id="name" value="<% if(locals.employer && employer.name != null) {%><%= employer.name %><% } %>" required="" class="form-control shadow bg-white rounded"/><br/>
+                    </div>
+
+                    <label class="control-label col-sm-8 pt-2 pb-1 m-0" for="surname">Nazwisko:</label>
+                    <div class="col-sm-8 mx-auto">
+                        <input type="text" id="surname" name="surname" rows="8" cols="64" class="form-control shadow bg-white rounded" value="<% if(locals.employer && employer.surname != null) {%><%= employer.surname %><% } %>" required="" class="form-control shadow bg-white rounded"/><br/>
```
Comment (macibol775) [inline views/employer/manage_employer.ejs:17]: rows cols??

File: views/organisation/manage_organisation.ejs [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+<%- include('../partials/navbar.ejs') %>
+<div class="container h-100">
+    <div class="row d-flex align-items-center justify-content-center h-100">
+        <div class="col-md-8 mx-auto text-center bg-white p-3">
+            <%if(!locals.organisation){%>
+                <h1 class="textTitle">Dodaj organizację</h1>
+            <%}else{%>
+                <h1 class="textTitle">Edytuj organizację</h1>
+            <%}%>
+            <form class="form-horizontal" id="offerForm">
```
Comment (macibol775) [inline views/organisation/manage_organisation.ejs:11]: not offer form

**Review by macibol775 — CHANGES_REQUESTED** [review]
Terrible html, ugly things in js, don't rely so much on copilot or copy-paste

File: controllers/adminController.js [hunk 19]
```diff
@@ -2,35 +2,166 @@ const Employer = require('../models/employerModel');
...
+        const employer_id = req.query._id;
+        const all_organisations = (await Organisation.find({})).map(org => org.toObject());
+        if (employer_id) {
+            const employer = (await Employer.findById(employer_id)).toObject();
+            const organisation = (await Organisation.findById(employer.organisation_id)).toObject();
+            res.render('employer/manage_employer', { title: 'Edycja konta pracowniczego', employer, organisation,all_organisations, user: req.session.employer, scrollable: true });
+        } else {
+            const passwordGenerator = ()=>{
+                const length = 8;
+                const charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*()_+?><:{}[]";
```
Comment (macibol775) [inline controllers/adminController.js:15]: = and - not allowed?

File: views/employer/manage_employer.ejs (hunk 12 above)
Comment (macibol775) [reply views/employer/manage_employer.ejs:19]: don't resolve unresolved issues!!!

File: controllers/adminController.js [hunk 20]
```diff
@@ -44,7 +44,7 @@ const addOrganisation_put = async (req, res) => {
     const { name, description } = req.body;
 
     try {
-        const org = await Organisation.create({ name, description });
+        Organisation.create({ name, description });
```
Comment (macibol775) [inline controllers/adminController.js:47]: BUG await

Comment (MasterCoookie) [reply controllers/adminController.js:47]: Yeah, this will not throw an error, but will also not work

File: models/employerModel.js [hunk 21]
```diff
@@ -31,7 +31,7 @@ const employerSchema = new mongoose.Schema({
     },
     organisation_id:{
         type: mongoose.Schema.Types.ObjectId,
-        required: [true, "Please provide an organisation"]
+        required: false
```
Comment (macibol775) [inline models/employerModel.js:34]: why?

Comment (MasterCoookie) [reply models/employerModel.js:34]: I requested it, tho, for future code, no need to specify not required

File: controllers/adminController.js [hunk 22]
```diff
@@ -2,35 +2,166 @@ const Employer = require('../models/employerModel');
...
 const modifyOrganistation_post = async (req, res) => {
+    const {name, description, org_id} = req.body;
+    try {
+        await Organisation.findByIdAndUpdate(org_id, { name, description});
+        console.log("Organisation %s modified", name);
+        res.sendStatus(201);
+    } catch (e) {
+        let errors = [];
+        if (e.errors) {
+            Object.values(e.errors).forEach(({ properties }) => {
```
Comment (tomazaw) [inline controllers/adminController.js:94]: syntax, if using one element in foreach ({}) is not needed

File: controllers/adminController.js [hunk 23]
```diff
@@ -2,35 +2,166 @@ const Employer = require('../models/employerModel');
 const Organisation = require('../models/organisationModel');
 // Should this be splitted into employerController and organisationController?
 const manageEmployer_get = async (req, res) => {
-
+    try {
+        const employer_id = req.query._id;
+        const all_organisations = (await Organisation.find({})).map(org => org.toObject());
```
Comment (tomazaw) [inline controllers/adminController.js:7]: i think that brackets inside find are not needed

File: public/js/modifyEmployer.js [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+                employerSaveResultDOMElement.classList.remove('alert-info');
+            } else {
+                employerSaveResultDOMElement.innerHTML = 'Błąd serwera!';
+                employerSaveResultDOMElement.classList.add('alert-danger');
+                employerSaveResultDOMElement.classList.remove('alert-info');
+            }
+        }
+    );
+})};
+export { initializeModifyEmployerSubmitListener as initializeModifyEmployerSubmitListener };
```
Comment (tomazaw) [inline public/js/modifyEmployer.js:35]: export as the same thing

File: views/admin/show_organisations.ejs [hunk 25]
```diff
@@ -0,0 +1,36 @@
+<%- include('../partials/head.ejs') %>
+<%- include('../partials/navbar.ejs') %>
+
+<div class="container bg-white" style="padding-top: 150px;">
+    <div class="row d-flex align-items-up justify-content-center h-100">
+        <h3 class="col-md-8 mx-auto text-center">Przeglądaj organizacje</h1>
```
Comment (tomazaw) [inline views/admin/show_organisations.ejs:6]: opened h3 closed h1

**Review by tomazaw — CHANGES_REQUESTED** [review]

**Review by macibol775 — APPROVED** [review]
ok

File: views/admin/show_organisations.ejs [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+                <tbody>
+                    <% organisations.forEach((organisation, index) => { %>
+                    <tr>
+                        <td><%= organisation.name %></td>
+                        <td>
+                            <a href="/admin/manage_organisation?_id=<%=organisation._id%>" class="btn btn-primary">Szczegóły</a>
+                        </td>
+                        <td>
+                            <form action="/admin/delete_organisation" method="post">
+                                <input type="hidden" name="_id" value="<%=organisation._id%>">
```
Comment (tomazaw) [inline views/admin/show_organisations.ejs:25]: no closing /

**Review by tomazaw — APPROVED** [review]

File: views/employer/manage_employer.ejs [hunk 27]
```diff
@@ -0,0 +1,62 @@
...
+                    <label class="control-label col-sm-8 pt-2 pb-1 m-0" for="password">Hasło:</label>
+                    <div class="col-sm-8 mx-auto">
+                        <input type="text" id="password" name="password" class="form-control shadow bg-white rounded" value="<%=passwordGenerator()%>" required="" class="form-control shadow bg-white rounded"/><br/>
+                    </div>
+                    <%}%>
+                </div>
+                <div class="col-sm-offset-2 col-sm-12">
+                    <button type="submit" class="btn" id="submitBtn">Opublikuj</button>
+                </div>
+                <div id="employerSaveResult" class="mt-2"></div>
```
Comment (macibol775) [inline views/employer/manage_employer.ejs:51]: not part of a form

File: views/organisation/manage_organisation.ejs [hunk 28]
```diff
@@ -0,0 +1,37 @@
...
+
+                    <label class="control-label col-sm-8 pt-2 pb-1 m-0" for="description">Opis organizacji:</label>
+                    <div class="col-sm-8 mx-auto">
+                        <textarea id="description" name="description" rows="8" cols="64" class="form-control shadow bg-white rounded"><% if(locals.organisation && organisation.description != null) {%><%= organisation.description %><% } %></textarea><br/>
+                    </div>
+                </div>
+                <div class="col-sm-offset-2 col-sm-12">
+                    <button type="submit" class="btn" id="submitBtn">Opublikuj</button>
+                </div>
+                <div id="organisationSaveResult" class="mt-2"></div>
```
Comment (macibol775) [inline views/organisation/manage_organisation.ejs:26]: not part of a form

**Review by macibol775 — CHANGES_REQUESTED** [review]
response message not part of a form

File: controllers/adminController.js (hunk 23 above)
Comment (MasterCoookie) [reply controllers/adminController.js:7]: Agreed

**Review by macibol775 — APPROVED** [review]
okk

File: controllers/adminController.js [hunk 29]
```diff
@@ -2,35 +2,166 @@ const Employer = require('../models/employerModel');
...
+        }
+        res.json({ errors });
+    }
 }
 
 const employer_delete = async (req, res) => {
-
+    const { _id } = req.body;
+    try {
+        await Employer.init();
```
Comment (MasterCoookie) [inline controllers/adminController.js:127]: Init most likely not needed

File: routes/adminRoutes.js
Comment (MasterCoookie) [inline routes/adminRoutes.js:1]: Protect routes soon

**Review by MasterCoookie — APPROVED** [review]
Check if init is needed and TODO - protect admin routes

**Pull request merged**